### PR TITLE
[Feature] Learning Goals

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Description

Every course has overarching themes/goals that students should acquire while doing the course. For Example: Learn HTML, or Learn CSS. This feature now allows instructors to define such course goals and connect them to corresponding lecture units. The learning goal progress is then calculated for a student from his/her performance in the connected lecture units (currently only exercises)

### Steps for Testing

Basically the test case would be to test the learning goals CRUD (Create, Read, Update,Delete) functionality from the instructor perspective (create learning goals, connect them to lecture units and so on) and check that the corresponding view from the student perspective is also updated

### Test Coverage

Server
- LearningGoalResource 82%
- LearningGoalService 95%

Client

- create-learning-goal 82%
- edit-learning-goal 95%
- learning-goal-form 89%
- learning-goal-detail-modal 91%
- learning-goal-management 86%
- course-learning-goals 96%
- learning-goal-popver 100%
- learning-goal-card 89%
- circular-progress-bar 100%

### Screenshots

**Instructors**
image
image

image
image
image
image
image
image
**Students**
image
image

### CODE REVIEW

**Review by marvinchimeka — APPROVED** [review]
tested the functionality on ts1. everything worked as it suppose to be. CRUD works on the instructor perspective as well as on the students perspective. approve

File: src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java [hunk 1]
```diff
@@ -17,16 +17,31 @@
...
+            WHERE lecture.course.id = :#{#courseId}
+            """)
+    Set<Lecture> findAllByCourseIdWithAttachments(@Param("courseId") Long courseId);
+
+    @Query("""
+                        SELECT lecture
+                        FROM Lecture lecture
+                        LEFT JOIN FETCH lecture.attachments
+                        LEFT JOIN FETCH lecture.lectureUnits
+                        WHERE lecture.course.id = :#{#courseId}
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java:33]: The indentation here looks off.

File: src/main/webapp/i18n/en/global.json [hunk 2]
```diff
@@ -299,5 +299,6 @@
         "connected": "Connected",
         "disconnected": "Disconnected"
     },
-    "loading": "Loading..."
+    "loading": "Loading...",
+    "completed": "Completed"
```
Comment (FrankeLukas) [inline src/main/webapp/i18n/en/global.json:303]: This appears to be missing in the `de` file.

File: src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<div class="progress-circle" [attr.data-percentage]="progressInPercent">
+    <span class="progress-left">
+        <span class="progress-bar-circle" [style.border-color]="circleColor"></span>
```
Comment (FrankeLukas) [inline src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.html:3]: So if I understand this correctly, Angular will call the `circleColor` function to get the color quite often (change detection?), even without `progressInPercent` changing. If so, wouldn't it make sense to cache that value in a field instead?

File: src/main/webapp/app/course/learning-goals/learning-goal-management/learning-goal-management.component.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        );
+    }
+
+    getLearningGoalProgress(learningGoal: LearningGoal) {
+        return this.learningGoalIdToLearningGoalProgress.get(learningGoal.id!);
+    }
+
+    loadData() {
+        this.isLoading = true;
+        this.learningGoalService
```
Comment (FrankeLukas) [inline src/main/webapp/app/course/learning-goals/learning-goal-management/learning-goal-management.component.ts:60]: This should probably use a resolver instead (which can then be reused for `course-learning-goals` too).

Comment (stefanwaldhauser) [reply src/main/webapp/app/course/learning-goals/learning-goal-management/learning-goal-management.component.ts:60]: No I am strictly against resolvers, they are a bad style imo

Comment (stefanwaldhauser) [reply src/main/webapp/app/course/learning-goals/learning-goal-management/learning-goal-management.component.ts:60]: image
https://blog.chinmoymohanty.com/2020/06/23/do-we-really-need-angular-route-resolvers/

Comment (FrankeLukas) [reply src/main/webapp/app/course/learning-goals/learning-goal-management/learning-goal-management.component.ts:60]: This was more about fetching the learning goals themselves, but it's not too important here so I won't insist. (Btw, you're setting `isLoading`, which looks like it's blocking to me, too.)

File: src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.html (hunk 3 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.html:3]: Good catch: 4f1d3097574eb0d550b45758427f47fce23b2d0d

File: src/main/webapp/i18n/en/global.json (hunk 2 above)
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/en/global.json:303]: Good catch! 9ca89cdfced9af0b6b1f3e7cfd1a7b842ce9634a

File: src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/repository/LectureRepository.java]: Good Catch! 09c53aa3fe1d67dfc7e1c2a114156a8c71725b90

**Review by airelawaleria — APPROVED** [review]
Everything works fine.

**Review by marvinchimeka — APPROVED** [review]
Tested on ts1. Like I described earlier everything worked as intended.

**Review by simonzilker — DISMISSED** [review]
Tested with @DenizShabani  on ts1. Worked as intended.

**Review by DenizShabani — DISMISSED** [review]
Tested on TS1 with @simonzilker and all works as intended.

**Review by JohannesWeiss — DISMISSED** [review]
Works as intended on ts1. Cool feature! 👍

**Review by LudwigWeide — APPROVED** [review]
Tested on TS3. Works as intended.

File: src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import { Component, Input, OnChanges, OnInit } from '@angular/core';
```
Comment (FrankeLukas) [inline src/main/webapp/app/shared/circular-progress-bar/circular-progress-bar.component.ts:1]: GitHub Actions: `'OnInit' is defined but never used`.

**Review by FrankeLukas — APPROVED** [review]
Works as advertised. Code looks good too on a quick glance (comments from this noon were addressed).

**Review by philgabr — DISMISSED** [review]
Tested in testing session. Everything works as intended.

**Review by valeriebucher — APPROVED** [review]
everything works on ts1. looks good!

**Review by EneaGore — APPROVED** [review]
Cool feature. Seems to all work fine.

**Review by tiffca — DISMISSED** [review]
All works as expected, create, view, update, deletion, view from the student side. Nice changes!

**Review by FlorianStang — DISMISSED** [review]
Everything workes as intended, cool Feature!

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        this.resultRepository = resultRepository;
+    }
+
+    /**
+     * Calculate the progress in a learning goal for a specific user
+     * @param learningGoal learning goal to get the progress for
+     * @param user user to get the progress for
+     * @return progress of the user in the learning goal
+     */
+    public LearningGoalProgress getLearningGoalPerformance(LearningGoal learningGoal, User user) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java:47]:
```suggestion
    public LearningGoalProgress calculateLearningGoalProgress(LearningGoal learningGoal, User user) {
```

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                return exercise.isAssessmentDueDateOver();
+            }
+            else {
+                return true;
+            }
+        }).collect(Collectors.toSet());
+
+        // In the case that two or more connected exercise units reference the same exercise, only one of them should be used in the progress calculation
+        Set<Exercise> exercisesAlreadyUsedInCalculation = new HashSet<>();
+
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java:68]: split this into 2 methods

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java [hunk 8]
```diff
@@ -77,11 +98,40 @@ public Lecture filterActiveAttachments(Lecture lectureWithAttachments, User user
 
     /**
      * Deletes the given lecture.
-     * Attachments are not explicitly deleted, as the delete operation is cascaded by the database.
+     * Attachments and Lecture Units are not explicitly deleted, as the delete operation is cascaded by the database.
      * @param lecture the lecture to be deleted
      */
     public void delete(Lecture lecture) {
-        lectureRepository.delete(lecture);
+        Optional<Lecture> lectureToDeleteOptional = lectureRepository.findByIdWithStudentQuestionsAndLectureUnitsAndLearningGoals(lecture.getId());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/LectureService.java:105]: please use Transactional here

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        for (LectureUnit lectureUnit : filteredLectureUnits) {
+            LearningGoalProgress.LectureUnitProgress lectureUnitProgress = new LearningGoalProgress.LectureUnitProgress();
+            lectureUnitProgress.lectureUnitId = lectureUnit.getId();
+            // ToDo implement way to track progress for lecture units other than exercise units
+            if (lectureUnit instanceof ExerciseUnit) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java:73]: this code is not very performant. I suggest to use the following existing queries:

```
// 1st: fetch participations, submissions and results for individual exercises
List<StudentParticipation> individualParticipations = participationService.findByStudentIdAndIndividualExercisesWithEagerSubmissionsResult(user.getId(),
        activeIndividualExercises);
// 2nd: fetch participations, submissions and results for team exercises
List<StudentParticipation> teamParticipations = participationService.findByStudentIdAndTeamExercisesWithEagerSubmissionsResult(user.getId(), activeTeamExercises);
// 3rd: merge both into one list for further processing
List<StudentParticipation> participations = Stream.concat(individualParticipations.stream(), teamParticipations.stream()).collect(Collectors.toList());
```

**Review by krusche — CHANGES_REQUESTED** [review]
Please improve the performance (as stated inline)

File: src/main/webapp/app/course/learning-goals/learning-goal-routing.module.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+const routes: Routes = [
+    {
+        path: ':courseId/goals/create',
+        component: CreateLearningGoalComponent,
+        data: {
+            authorities: ['ROLE_ADMIN', 'ROLE_INSTRUCTOR'],
+            pageTitle: 'artemisApp.learningGoal.createLearningGoal.title',
+        },
+        canActivate: [UserRouteAccessService],
```
Comment (FrankeLukas) [inline src/main/webapp/app/course/learning-goals/learning-goal-routing.module.ts:16]: This will need an update to properly display the human readable breadcrumbs since #2474 was merged.

**Comment by FrankeLukas** [comment]
The breadcrumbs are adjusted now (`AnotherTest` is my course name):
grafik
grafik
grafik

I also fixed the GitHub Actions warning while I was at it.

**Comment by stefanwaldhauser** [comment]
> The breadcrumbs are adjusted now (`AnotherTest` is my course name):
> grafik
> grafik
> grafik
> 
> I also fixed the GitHub Actions warning while I was at it.

Thx so much :)

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java (hunk 8 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LectureService.java:108]: Done in ae8bec507791f3f824dd5125a53b82621e651ead

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java (hunk 6 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java]: Done in da86f2c21800028e9a9a960cd727cf04ede905c7

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java (hunk 7 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java]: Done in da86f2c21800028e9a9a960cd727cf04ede905c7

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java]: Done in da86f2c21800028e9a9a960cd727cf04ede905c7

**Comment by stefanwaldhauser** [comment]
> Please improve the performance (as stated inline)

Done in da86f2c21800028e9a9a960cd727cf04ede905c7

**Review by JohannesWeiss — APPROVED** [review]
Re-Tested on ts3 and worked as intended.

Just a small visual improvement could be made by increasing the gap between the warning for an empty title and the input field.
Screenshot 2020-12-10 at 18 45 41

**Review by airelawaleria — APPROVED** [review]
Re-tested on ts3. Works fine

**Review by simonzilker — APPROVED** [review]
Tested with @airelawaleria on ts3. Works as intended!

**Review by JannisBittlmayer — DISMISSED** [review]
Tested on TS3, works as intended

**Review by LudwigWeide — APPROVED** [review]
Tested on TS1. Works as intended.

**Review by DenizShabani — DISMISSED** [review]
Tested on TS3 and works as intended

**Review by JohannesWeiss — COMMENTED** [review]
The achieved points seem to be displayed with some kind of floating point error:
image
The results view shows 2.5 points, which is the correct value. Tested on ts1 with an existing exercise, but with a new participation.

**Review by tiffca — DISMISSED** [review]
Tested functionality successfully on test server 1 (create/delete/view as a student/values update accordingly). Noticed the button moved to the course details page.

**Review by airelawaleria — COMMENTED** [review]
Same issue as @JohannesWeiss .
Also updating the exercises counted for a certain learning goal results in an error.
image

**Comment by krusche** [comment]
@stefanwaldhauser please fix the 500 found by @airelawaleria 
Also double check that the tests are running properly and that the merged changes from `develop` do not lead to any issues

**Comment by krusche** [comment]
sorry, the latest merge broke your PR

**Comment by stefanwaldhauser** [comment]
> Same issue as @JohannesWeiss .
> Also updating the exercises counted for a certain learning goal results in an error.
> image
@airelawaleria 
Could you maybe give me more explanation please as I could not reproduce it locally :)

**Comment by stefanwaldhauser** [comment]
Server Test fails because of flaky test unrelated to this PR: 
image

**Comment by stefanwaldhauser** [comment]
> Same issue as @JohannesWeiss .
> Also updating the exercises counted for a certain learning goal results in an error.
> image

should be fixed in b100691dd427aeb5a565bef8595080df3dc2ac52

**Comment by stefanwaldhauser** [comment]
> The achieved points seem to be displayed with some kind of floating point error:
> image
> The results view shows 2.5 points, which is the correct value. Tested on ts1 with an existing exercise, but with a new participation.

should be fixed in https://github.com/ls1intum/Artemis/pull/2510#pullrequestreview-550385694

If not please tell me the exact exercise and account where this happens so I can recreate it (What value gets rounded to the wrong number and so on)

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java [hunk 11]
```diff
@@ -25,16 +25,20 @@
 
     @ManyToOne
     @JoinColumn(name = "course_id")
+    @JsonIgnoreProperties("learningGoals")
     private Course course;
 
+    // ToDo remove this connection, only go through lecture units
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java:31]: is this TODO for a later PR?

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 12]
```diff
@@ -88,12 +88,25 @@
...
+    List<StudentParticipation> findByExerciseIdAndStudentId(@Param("exerciseId") Long exerciseId, @Param("studentId") Long studentId);
+
     @Query("select distinct participation from StudentParticipation participation left join fetch participation.results left join fetch participation.submissions where participation.exercise.id = :#{#exerciseId} and participation.student.id = :#{#studentId}")
     List<StudentParticipation> findByExerciseIdAndStudentIdWithEagerResultsAndSubmissions(@Param("exerciseId") Long exerciseId, @Param("studentId") Long studentId);
 
     @Query("select distinct participation from StudentParticipation participation left join fetch participation.submissions where participation.exercise.id = :#{#exerciseId} and participation.team.id = :#{#teamId}")
     List<StudentParticipation> findByExerciseIdAndTeamIdWithEagerSubmissions(@Param("exerciseId") Long exerciseId, @Param("teamId") Long teamId);
 
+    @Query("""
+            select distinct participation
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:104]: it would be nice to have a consistent writing style: In other queries your are using SELECT in uppercase (which looks way more beautiful ;) )

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+     * progress by re-trying a quiz as often as you like. It is therefore normal, that the points here might differ from the points officially achieved in an exercise.
+     *
+     * @param exerciseUnits exercise units to check
+     * @param user user to check for
+     * @return progress of the user in the exercise units
+     */
+    public Set<LearningGoalProgress.LectureUnitProgress> calculateExerciseUnitsProgress(Set<ExerciseUnit> exerciseUnits, User user) {
+        // for each exercise unit, the exercise will be mapped to a freshly created lecture unit progress.
+        Map<Exercise, LearningGoalProgress.LectureUnitProgress> exerciseToLectureUnitProgress = exerciseUnits.stream().filter(exerciseUnit -> exerciseUnit.getExercise() != null)
+                .filter(exerciseUnit -> exerciseUnit.getExercise().isAssessmentDueDateOver()).collect(Collectors.toMap(ExerciseUnit::getExercise, exerciseUnit -> {
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java:41]: here you could combine both filters into one:  

.filter(exerciseUnit -> exerciseUnit.getExercise() != null && exerciseUnit.getExercise().isAssessmentDueDateOver()).collect(...

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+            else {
+                Submission latestSubmission = latestSubmissionOptional.get();
+                Result latestResult = latestSubmission.getLatestResult();
+                if (latestResult == null) {
+                    return Optional.empty();
+                }
+                else {
+                    return Optional.of(latestResult);
+                }
+
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java:106]:
```suggestion
```

File: src/main/webapp/i18n/de/learningGoal.json [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                "title": "Lernzielmanagement"
+            },
+            "createLearningGoal": {
+                "title": "Erstelle ein neues Lernziel",
+                "titlePlaceholder": "Gib dem Lernziel einen Titel",
+                "titleRequiredValidationError": "Ein Lernziel braucht einen Titel",
+                "titleMaxLengthValidationError": "Der Titel eines Lernziels ist auf maximal 255 Symbole beschraenkt",
+                "titleUniqueValidationError": "Es gibt bereits ein Lernziel mit diesem Titel in dem Kurs",
+                "descriptionPlaceholder": "Beschreibe das Lernziel. Was sollen die Studenten lernen?",
+                "descriptionMaxLengthValidationError": "Die Beschreibung eines Lernziels ist auf maximal 10 000 Symbole beschraenkt",
```
Comment (derLalla) [inline src/main/webapp/i18n/de/learningGoal.json:20]:
```suggestion
                "titleMaxLengthValidationError": "Der Titel eines Lernziels ist auf maximal 255 Symbole beschränkt",
                "titleUniqueValidationError": "Es gibt bereits ein Lernziel mit diesem Titel in dem Kurs",
                "descriptionPlaceholder": "Beschreibe das Lernziel. Was sollen die Studenten lernen?",
                "descriptionMaxLengthValidationError": "Die Beschreibung eines Lernziels ist auf maximal 10 000 Symbole beschränkt",
```

**Review by derLalla — CHANGES_REQUESTED** [review]
I've looked throughly through your code and it looks very good! I've only remarked some smaller issues.
The branch was not compiling, so I couldn't test it on the testserver :/

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java (hunk 11 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/LearningGoal.java]: Done in : c70aa0f01060fd36a9f540c34c4b5f0e009a4410

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java (hunk 12 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java]: c70aa0f01060fd36a9f540c34c4b5f0e009a4410

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java (hunk 13 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java]: c70aa0f01060fd36a9f540c34c4b5f0e009a4410

File: src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java (hunk 14 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java]: c70aa0f01060fd36a9f540c34c4b5f0e009a4410

File: src/main/webapp/i18n/de/learningGoal.json (hunk 15 above)
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/de/learningGoal.json]: c70aa0f01060fd36a9f540c34c4b5f0e009a4410

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
There still seems to be some kind of floating point error in the overview.
Screenshot 2020-12-14 at 23 16 05
1.5 points would be correct.
Tested on ts3 with @marvinchimeka , @airelawaleria  and @simonzilker .

**Review by airelawaleria — COMMENTED** [review]
Tested in the testing session. Confirm the issue above.

**Review by simonzilker — COMMENTED** [review]
Tested in testing session.  Can confirm the issue.

**Review by marvinchimeka — COMMENTED** [review]
tested it in the testing session. I can confirm this issue.

**Review by derLalla — CHANGES_REQUESTED** [review]
it would be nice, if the dropdown menu is either disabled or shows an alert if no lectures are available. I had no Lectures and the dropdown was therefore empty. 

image

**Comment by stefanwaldhauser** [comment]
> There still seems to be some kind of floating point error in the overview.
> Screenshot 2020-12-14 at 23 16 05
> 1.5 points would be correct.
> Tested on ts3 with @marvinchimeka , @airelawaleria and @simonzilker .

Replaced the old UI with a new one that is hopefully more clearer: ed136261eda9fb7980b4427696dcd198fe37b8b6
image

**Comment by stefanwaldhauser** [comment]
> it would be nice, if the dropdown menu is either disabled or shows an alert if no lectures are available. I had no Lectures and the dropdown was therefore empty.
> 
> image

artemisApp.learningGoal.createLearningGoal.noLectures

**Review by airelawaleria — APPROVED** [review]
Re-tested with @simonzilker, @JohannesWeiss and @marvinchimeka .Works fine 👍

**Review by simonzilker — APPROVED** [review]
tested in testing session. Re-approve!

**Review by JohannesWeiss — APPROVED** [review]
Re-Tested on ts3 in testing session wit @simonzilker, @marvinchimeka and @airelawaleria. The floating point problems are nicely solved, empty lectures also work as intended. Good job. 👍

**Review by marvinchimeka — APPROVED** [review]
tested in a testing session. Approve

**Review by tiffca — APPROVED** [review]
Latest changes look good, the view when there isn't a lecture and the new learning goals details page. Nice changes!

**Review by derLalla — APPROVED** [review]
re-approve. It's a really nice feature! :)

**Review by JannisBittlmayer — APPROVED** [review]
Tested on TS3, works really well and as intended

**Review by krusche — APPROVED** [review]
